### ISSUE TITLE
Learning rate schedules do not affect batchnorm

### ISSUE THREAD

**Issue by ndryden** [LLNL/lbann#98]
Batchnorm (and potentially other similar regularizers) has a two optimizers to learn its scale and shift parameters. However, since it's a regularizer and not a learning layer, our learning rate schedules are not applied to it.

We could make it a learning layer, but this adds unneeded things like weights to it, and learning layers assume they have only one optimizer.

**Comment by ndryden** [comment]
Some recent changes in batchnorm have made the relatively easier to handle: all the parameters are kept within a single `m_parameters` (gradients in `m_parameters_gradient`) matrix that is managed with views. It's therefore possible to switch to having only one optimizer.

Further, this format then maps precisely onto a learning layer, where we keep the parameters in the weights matrix and use the provided optimizer.

This gives two choices:
1. Make batchnorm only a subclass of `learning` (easy-- currently `regularizer_layer` provides nothing).
2. Keep the `regularizer_layer` parent and add `learning`. This introduces diamond inheritance, which would require us to make `regularizer_layer` and `learning` inherit virtually from `Layer` and will introduce issues with constructors, etc.

My thought is to take choice 1. We lose the semantic meaning of being a regularizer (not that anything relies on that, IIRC) but it simplifies code. What does everyone else think?

(A more radical choice would be to turn `regularizer_layer` (and possibly `transform_layer`) into interfaces but that would probably involve rethinking much of our object hierarchy again.)

**Comment by davidHysom** [comment]
I'm in favor of 1.

How, if at all, will this affect the prototext side? Asked a different way,

are there any changes to the batchnorm layer constructors?

________________________________

**Comment by ndryden** [comment]
Discussed with Brian and leaned toward a variant of 1, creating a new `learning_regularizer` layer. This would change the batchnorm constructor, to add an optimizer parameter. (I took care of doing this.)

I ended up also adding an ABC `optimizable_layer` that defines a `get_optimizer` method as a pure virtual (think of this as an interface). `learning` and `learning_regularizer` now also inherit from `optimizable_layer`. Note this class doesn't define any data members, to avoid issues with multiple inheritance.
